### ISSUE TITLE
HIP_VISIBLE_DEVICES required on multi-GPU systems for correct output

### ISSUE THREAD

**Issue by MarkusBuettner** [AdaptiveCpp/AdaptiveCpp#1180]
**Bug summary**
On a system with an AMD MI 210 and a MI 100 GPU, a simple test code provides wrong results if HIP_VISIBLE_DEVICES is not set.

**To Reproduce**
Take this test code:
```c++
#include <sycl/sycl.hpp>
#include <iostream>

int main() {
        sycl::queue queue;

        std::cout << "SYCL Device name: " << queue.get_device().get_info<sycl::info::device::name>() << "\n";

        constexpr size_t len = 1024 * 1024;

        sycl::buffer<float, 1> a{len};
        sycl::buffer<float, 1> b{len};

        queue.submit([&](sycl::handler& cgh) {
                auto aAcc = a.get_access<sycl::access_mode::discard_write>(cgh);
                auto bAcc = b.get_access<sycl::access_mode::discard_write>(cgh);
                cgh.parallel_for(len, [=](sycl::id<1> id) {
                        aAcc[id] = 1.0f;
                        bAcc[id] = 2.0f;
                });
        }).wait();

        auto aAcc = a.get_host_access();
        auto bAcc = b.get_host_access();
        std::cout << aAcc[0] << " " << bAcc[0] << "\n";

        return 0;
}
```
I compiled it with `$WORK/acpp-rocm/bin/acpp --acpp-targets="omp;hip:gfx90a" -g hello_sycl.cpp -o hello_sycl_acpp`
and now, I can observe the following behaviour:
```
~$ ./hello_sycl_acpp 
SYCL Device name: AMD Instinct MI210
-0.372549 -0.372549
~$ HIP_VISIBLE_DEVICES=0 ./hello_sycl_acpp 
SYCL Device name: AMD Instinct MI210
1 2
~$ HIP_VISIBLE_DEVICES=1 ./hello_sycl_acpp 
SYCL Device name: AMD Instinct MI100
-0.372549 -0.372549
~$ ACPP_VISIBILITY_MASK=omp ./hello_sycl_acpp 
SYCL Device name: hipSYCL OpenMP host device
1 2
```

If I compile with target hip:gfx908 (for MI 100), I can see a similar output, except that I get the correct output only for HIP_VISIBLE_DEVICES=1

**Expected behavior**
Honestly, I'm not sure what the "correct" behaviour here is. AFAIR with Intel oneAPI, Code compiled for gfx90a can only run on gfx90a devices, otherwise it fails with an error message.
But if the code is compiled for gfx90a (MI 210), and the queue binds to the MI 210, I would expect correct results.

**Describe your setup**
I installed AdaptiveCpp this morning by pulling the development branch and building against ROCm 5.6.0 LLVM

**Optional additional diagnostic information**
I attached the output of ACPP_DEBUG_LEVEL=3 if HIP_VISIBLE_DEVICES=0 is set ([debug_log_env.txt](https://github.com/AdaptiveCpp/AdaptiveCpp/files/12857430/debug_log_env.txt))
and if it is not set ([debug_log_noenv.txt](https://github.com/AdaptiveCpp/AdaptiveCpp/files/12857432/debug_log_noenv.txt))

*Edit:*
It seems to be a problem with ROCm rather than this compiler. If I take [AMD's HIP Hello World Code](https://github.com/ROCm-Developer-Tools/HIP-Examples/blob/master/HIP-Examples-Applications/HelloWorld/HelloWorld.cpp)  and compile it with `hipcc --amdgpu-target=gfx90a hip-helloworld.cpp -o hip_helloworld` I see similar behaviour as descirbed above.

So maybe a hint somewhere (maybe at queue construction?)  could be helpful.

**Comment by illuhad** [comment]
ROCm does not like not having binaries for all GPUs in a multi-GPU system. Please try `--acpp-targets="omp;hip:gfx908,gfx90a"`.
As you point out this is a ROCm limitation. We also cannot easily check for this problem at the moment to issue a warning.

For such use cases though, you are deep in the realm of what SSCP is designed to do. So the AdaptiveCpp answer to this problem is to just set `--acpp-targets=generic` which should also resolve your issue.

**Comment by MarkusBuettner** [comment]
With `--acpp-targets="omp;hip:gfx908,gfx90a"` it works as intended. I was just wondering, because I thought that with oneAPI and specifying only one of gfx908 / gfx90a I get an error if I try to run it on the other architecture. I have to try that again, though. 

Regarding SSCP: Yes, that what I want to do - but building LLVM takes some time...

**Comment by illuhad** [comment]
> With --acpp-targets="omp;hip:gfx908,gfx90a" it works as intended. I was just wondering, because I thought that with oneAPI and specifying only one of gfx908 / gfx90a I get an error if I try to run it on the other architecture. I have to try that again, though.

With `--acpp-targets=hip`, you are using a compilation flow focusing on interoperability; it just reuses the clang HIP frontend that hipcc uses too. This means that kernel binary handling is exactly the same as in ROCm/hipcc, for better or worse. Depending on how oneAPI uses the HIP API, perhaps they use different kernel binary injection methods compared to what the clang HIP toolchain autogenerates. Similarly SSCP handles things quite differently. So, long story short, there might be differences between oneAPI, AdaptiveCpp HIP target, and AdaptiveCpp SSCP target.

> Regarding SSCP: Yes, that what I want to do - but building LLVM takes some time...

`apt.llvm.org` (or distribution repositories) is your friend ;) Of course for the HIP target you need LLVM too.

**Comment by MarkusBuettner** [comment]
The machine I'm testing this on is part of a cluster, so I used spack to install LLVM. And it took only 40 minutes :-)

With the SSCP target, I get an undefined symbol "hipDriverGetVersion" error in librt-backend-hip.so. 
librt-backend-hip.so is linked against "libhiprtc.so.5", but hipDriverGetVersion is only defined in libamdhip64.so. 

If I manually preload that library, the example code works, but our model code produces NaNs somewhere, which doesn't happen with the ahead-of-time compilation...

**Comment by illuhad** [comment]
> With the SSCP target, I get an undefined symbol "hipDriverGetVersion" error in librt-backend-hip.so.
librt-backend-hip.so is linked against "libhiprtc.so.5", but hipDriverGetVersion is only defined in libamdhip64.so.

We do tell cmake to link the HIP backend plugin against `libamdhip64`, but for whatever reason on some systems this does not work and I don't know yet why. With `--acpp-targets=hip` on the other hand `acpp` links the generated binary with `-lamdhip64`, so you don't get this issue even though the same problem is in principle there as well.
Preloading the library is the right workaround at the moment :)

>  but our model code produces NaNs somewhere, which doesn't happen with the ahead-of-time compilation...

The difference is not just JIT vs ahead-of-time, it's an entirely different compiler. The SSCP compiler does not yet support all the functionality of the old compiler (e.g. SYCL 2020 group algorithms) but mostly you get a JIT failure if you hit one of these cases. I would recommend investigating what is going on there, as it could be some UB in your code, or of course a compiler bug on our side. In the latter case it would be a high priority issue because our compiler work in the future is going to focus almost exclusively on SSCP. If you have a reproducer, I'd love to look at it :-)

**Comment by MarkusBuettner** [comment]
Well I can say that it also happens for CUDA JIT compilation. Here I'm also getting some messages:
```
'+ptx86' is not a recognized feature for this target (ignoring feature)
'+ptx86' is not a recognized feature for this target (ignoring feature)
'+ptx86' is not a recognized feature for this target (ignoring feature)
```
(Tested on an RTX 3070 with CUDA 12.2 and driver 535.104.12)

Is there a list of which features are (not) supported with SSCP? The code contains just two `nd_range<1>` kernels, where one of them uses a local accessor. Apart from that, we are not using group algorithms and I'm not aware of any specific places where there could be UB. And the code runs with JIT on CPUs, but not multi-threaded. 

I can try to create a reproducer, I guess you don't want to dig through our complete code to figure it out :-)

Edit: Also compute-sanitizer does not report any errors, but I'm not sure if that means anything...

**Comment by illuhad** [comment]
> '+ptx86' is not a recognized feature for this target (ignoring feature)
'+ptx86' is not a recognized feature for this target (ignoring feature)
'+ptx86' is not a recognized feature for this target (ignoring feature)

This is expected with CUDA SSCP on more recent GPUs at the moment. It's not a problem per se and can be ignored.

> Is there a list of which features are (not) supported with SSCP?

Unsupported is really only group algorithms except for `group_barrier`, `sycl::reduction` and some extensions like `hipSYCL_enqueue_custom_operation` and scoped parallelism. `nd_range` and local memory is supported.

> And the code runs with JIT on CPUs, but not multi-threaded.

SSCP does not support CPU yet. If you have `--acpp-targets=generic` and then run on the CPU, you will run through the ahead-of-time OpenMP backend, except that there will not be `#pragma omp parallel for` unless you use `--acpp-targets="omp;generic"`.

> I can try to create a reproducer, I guess you don't want to dig through our complete code to figure it out :-)

Thanks, that would be great :)

> Edit: Also compute-sanitizer does not report any errors, but I'm not sure if that means anything...

Yeah.. perhaps it's some builtin function call that is not mapped correctly or so.

EDIT: Which LLVM version are you using?

**Comment by MarkusBuettner** [comment]
I've tested this with LLVM versions 14 and 16, with both having this issue. 

I'm also still trying to find a reproducer, but everything I tried so far worked with SSCP... Maybe my examples were too simple.

**Comment by illuhad** [comment]
Maybe also check the host code; perhaps there is a synchronization issue or so and it just happens to work without SSCP due to different kernel timings.

**Comment by MarkusBuettner** [comment]
I took a different approach and ran both, the ahead-of-time compiled and the SSCP program through Nsight Compute profiler. And here I found something curious:

We have two kernels, one small one which issued in my test around 8800 instructions for both the ahead-of-time code and SSCP. And from the SASS code it looks like this one is translated correctly.

The bigger one issues in the AOT program 190.000 - 200.000 instructions, in the SSCP case 7.000. 
This kernel looks roughly like this:
```c++
cgh.parallel_for(sycl::nd_range<1>{ {num_elems_padded}, { local_group_size } }, 
  [=](sycl::nd_item<1> tid) {
    size_t el_id = tid.get_global_id(0);
    if (el_id >= num_elems) return;

    element_local[tid.get_local_id(0)] = element[el_id]; // This just loads some struct into shared memory

    integration::process_element(...); // I omitted all the parameters -- this is a templated function from a header file
}
```
Now if I look at the SASS code, I can still see the global load, shared store; but what should come next in the integration function appears to be missing, the only thing I can then see is some global stores before the EXIT statement. 

So maybe it has an issue with compiling the templated code? However, I failed to extract a minimal example where this also happens. I'll check whether we can give you access to our code, maybe it is some very specific edge-case that we happened to trigger.

**Comment by illuhad** [comment]
> So maybe it has an issue with compiling the templated code? However, I failed to extract a minimal example where this also happens. I'll check whether we can give you access to our code, maybe it is some very specific edge-case that we happened to trigger.

Do you perhaps use `__SYCL_DEVICE_ONLY__` macro (or other mechanism?) to specialize code for host and device? Because SSCP uses a single-pass compiler, this macro is never defined for SSCP, and indeed the SYCL specification only defines it as an implementation detail for multipass compilers, not as a portable mechanism for code specialization across multiple implementation choices.

SSCP operates almost entirely at the level of LLVM IR, so all the template logic is handled by the clang frontend in unchanged manner as for the host code.

Most likely there's no need to look at SASS, let's stick to the LLVM level first :-) If you compile with `-mllvm -hipsycl-sscp-emit-hcf` for the translation unit with the problematic kernel, the compiler will dump the HCF file in the source directory. The HCF file contains the generated device LLVM IR along with metadata. We can extract the LLVM IR using `acpp-hcf-tool`. Can you please provide the device IR (or the HCF, then I can extract it myself)?

**Comment by MarkusBuettner** [comment]
Thank's for taking the time to look at it :) I've extracted the hcf file:
[hcf_file.zip](https://github.com/AdaptiveCpp/AdaptiveCpp/files/12873020/hcf_file.zip)

There are no device-specific macros in there, we try to avoid that for now.

**Comment by illuhad** [comment]
Thanks, the original IR contains a lot of code for the functions in question. Right before translating to PTX, this is what it generates for me as optimized IR for CUDA:
https://godbolt.org/z/rrzMb9TWz

Here the `process_element` block indeed does no longer contain a lot of arithmetic, however, the kernel still does, in particular I see lots of calls to special functions. SSCP inlines very aggressively (more than the other compilation flows), so most likely it has just reordered instructions after inlining so that `process_element` itself got slimmed down.

So far I see no evidence so far that something went wrong, no undefs creeping into the code or other clear compiler problems.

**Comment by MarkusBuettner** [comment]
Is there a way to control how aggressively the compiler inlines/optimizes code? Like how you would -O0/1/2/3 to clang. 

I would understand if the compiler can optimize some instructions away, but as I said it reduces from 180.000 instructions to 7.000 - a reduction by 94 % is in my opinion suspiciously high.

**Comment by illuhad** [comment]
It's JIT... it always optimizes.

> I would understand if the compiler can optimize some instructions away, but as I said it reduces from 180.000 instructions to 7.000 - a reduction by 94 % is in my opinion suspiciously high.

Are you talking about executed instructions or binary size? In the case of the former, I would recommend to also compare binary size to that which the other compiler flows generate. Perhaps this not due to compiler optimizations but due to runtime arguments which then cause the kernel to enter other branches. From the IR I have seen, I imagine a lot of these instructions are generated by calls to special functions, and the amount of computation there can depend greatly on input values.

**Comment by MarkusBuettner** [comment]
These are the executed instructions as recorded by the Nsight Compute Profiler, not binary size. Execution is mostly uniform across all elements, so there shouldn't be that much deviation between the number of instructions. 

I've also sent you an email with some more background information and the actual code.

**Comment by MarkusBuettner** [comment]
I think this issue can be closed. With SSCP, HIP_VISIBLE_DEVICES is no longer needed; for the "classical" compilation with `--acpp-targets` HIP_VISIBLE_DEVICES is needed, but this is not specific to AdaptiveCpp.

The remaining issues with SSCP have been resolved with PR #1187 and some subsequent errors with #1191, #1192, #1193
